go/vt/discovery/healthcheck.go

```diff
@@ -412,44 +395,74 @@ func (hc *HealthCheckImpl) ServeHTTP(w http.ResponseWriter, _ *http.Request) {
 // servingConnStats returns the number of serving tablets per keyspace/shard/tablet type.
 func (hc *HealthCheckImpl) servingConnStats() map[string]int64 {
 	res := make(map[string]int64)
-	hc.mu.RLock()
-	defer hc.mu.RUnlock()
-	for _, hcc := range hc.addrToConns {
-		hcc.mu.RLock()
-		if !hcc.tabletStats.Up || !hcc.tabletStats.Serving || hcc.tabletStats.LastError != nil {
-			hcc.mu.RUnlock()
+	hc.mu.Lock()
+	defer hc.mu.Unlock()
+	for _, th := range hc.addrToHealth {
+		if !th.tabletStats.Up || !th.tabletStats.Serving || th.tabletStats.LastError != nil {
 			continue
 		}
-		key := fmt.Sprintf("%s.%s.%s", hcc.tabletStats.Target.Keyspace, hcc.tabletStats.Target.Shard, topoproto.TabletTypeLString(hcc.tabletStats.Target.TabletType))
-		hcc.mu.RUnlock()
+		key := fmt.Sprintf("%s.%s.%s", th.tabletStats.Target.Keyspace, th.tabletStats.Target.Shard, topoproto.TabletTypeLString(th.tabletStats.Target.TabletType))
 		res[key]++
 	}
 	return res
 }
 
+// updateHealth updates the tabletHealth record and transmits the tablet stats
+// to the listener.
+func (hc *HealthCheckImpl) updateHealth(ts *TabletStats, conn queryservice.QueryService) {
+	// Unconditionally send the received update at the end.
+	defer func() {
+		if hc.listener != nil {
+			hc.listener.StatsUpdate(ts)
+		}
+	}()
+
+	hc.mu.Lock()
+	th, ok := hc.addrToHealth[ts.Key]
+	if !ok {
+		// This can happen on delete because the entry is removed first.
+		hc.mu.Unlock()
+		return
+	}
+	oldts := th.tabletStats
+	th.tabletStats = *ts
+	th.conn = conn
+	hc.mu.Unlock()
+
+	// In the case where a tablet changes type (but not for the
+	// initial message), we want to log it, and maybe advertise it too.
+	if oldts.Target.TabletType != topodatapb.TabletType_UNKNOWN && oldts.Target.TabletType != ts.Target.TabletType {
+		// Log and maybe notify
+		log.Infof("HealthCheckUpdate(Type Change): %v, tablet: %s, target %+v => %+v, reparent time: %v",
+			oldts.Name, topotools.TabletIdent(oldts.Tablet), topotools.TargetIdent(oldts.Target), topotools.TargetIdent(ts.Target), ts.TabletExternallyReparentedTimestamp)
+		if hc.listener != nil && hc.sendDownEvents {
+			oldts.Up = false
+			hc.listener.StatsUpdate(&oldts)
+		}
+
+		// Track how often a tablet gets promoted to master. It is used for
+		// comparing against the variables in go/vtgate/buffer/variables.go.
+		if oldts.Target.TabletType != topodatapb.TabletType_MASTER && ts.Target.TabletType == topodatapb.TabletType_MASTER {
+			hcMasterPromotedCounters.Add([]string{ts.Target.Keyspace, ts.Target.Shard}, 1)
+		}
+	}
+}
+
 // finalizeConn closes the health checking connection and sends the final
 // notification about the tablet to downstream. To be called only on exit from
 // checkConn().
 func (hc *HealthCheckImpl) finalizeConn(hcc *healthCheckConn) {
-	hcc.mu.Lock()
-	hccConn := hcc.conn
-	hccCtx := hcc.ctx
-	hcc.conn = nil
+	if hcc.conn != nil {
+		// Don't use hcc.ctx because it's already closed.
+		hcc.conn.Close(context.Background())
```

Comment:
As we found out the hard way, under certain circumstances this could take a long time. so I think it's cleaner to update the stats to mark the connection as not serving _before_ doing the connection close which is how I did it before.